src/grader_webhook.ts

```diff
@@ -0,0 +1,415 @@
+import {decode, getRandomId} from './util';
+import {findSubmissionById} from './submissions';
+import {pipe} from 'fp-ts/function';
+import * as D from 'io-ts/Decoder';
+import {JwesDecoder} from './crypto/jwes_decoder';
+import * as Db from './db';
+import {TaskSubtask, TaskTest} from './db_models';
+import {findTaskById} from './tasks';
+import {longPollingHandler} from './long_polling';
+
+export const taskGraderWebhookPayloadDecoder = pipe(
+  D.struct({
+    sToken: D.string,
+  }),
+);
+export type TaskGraderWebhookPayload = D.TypeOf<typeof taskGraderWebhookPayloadDecoder>;
+
+export interface ProgramExecutionResult {
+  noFeedback: boolean,
+  exitSig: number,
+  exitCode: number,
+  continueOnError?: boolean,
+  stdout: {
+    data: string,
+  },
+  stderr: {
+    data: string,
+  },
+  files: any,
+  timeTakenMs: number,
+  memoryUsedKb: number,
+}
+
+export interface GraderResultExecution {
+  id: number,
+  name: string,
+  testsReports: {
+    name: string,
+    execution?: ProgramExecutionResult,
+    checker?: ProgramExecutionResult,
+    extraChecker?: ProgramExecutionResult,
+    sanitizer: ProgramExecutionResult,
+  }[],
+}
+
+export interface TokenParams {
+  sTaskName: string,
+  sResultData: {
+    executions: GraderResultExecution[],
+    solutions: {
+      compilationExecution: ProgramExecutionResult,
+    }[],
+  },
+}
+
+async function createNewTest(idSubmission: string, idTask: string, testName: string): Promise<TaskTest> {
+  let maxRank = await Db.querySingleScalarResult<number>(`SELECT MAX(tm_tasks_tests.iRank) from tm_tasks_tests
+  JOIN tm_submissions ON tm_submissions.idTask = tm_tasks_tests.idTask
+  WHERE tm_submissions.ID = :idSubmission and tm_tasks_tests.sGroupType = 'Evaluation';`, {
+    idSubmission,
+  });
+
+  if (null === maxRank) {
+    maxRank = 0;
+  }
+
+  const ID = getRandomId();
+
+  await Db.execute('INSERT INTO tm_tasks_tests (ID, idTask, sGroupType, iRank, bActive, sName) values (:ID, :idTask, \'Evaluation\', :iRank, 1, :sName)', {
+    ID,
+    idTask,
+    iRank: maxRank + 1,
+    sName: testName,
+  });
+
+  return {
+    ID,
+    idTask,
+    idSubmission,
+    sName: testName,
+    sGroupType: 'Evaluation',
+    iRank: maxRank + 1,
+    sOutput: '',
+    idSubtask: null,
+    idUser: null,
+    idPlatform: null,
+    bActive: 1,
+    sInput: '',
+    iVersion: 0,
+  };
+}
+
+export async function receiveSubmissionResultsFromTaskGrader(taskGraderWebhookPayload: unknown): Promise<void> {
```

Comment:
I agree, this function is a transcription of the code in this file: https://github.com/France-ioi/TaskPlatform/blob/46916f84ec7023af894b2b3230ed9d888fa4ba1f/grader/pushResults.php
My problem is that this code is indeed difficult to read and I didn't work to break it while doing the porting from TaskPlatform to CodecastTaskBackend, so I've kept it as similar as possible to the orginal.

Reply:
> > My problem is that this code is indeed difficult to read and I didn't work to break it while doing the porting from TaskPlatform to CodecastTaskBackend,
> 
> If you had tests it would be much easier

The TaskPlatform project doen't have any test unfortunately